CI on GKE: Trigger PR Jenkins based on webhooks.

spxtr opened the pull request:
Sorry about the huge PR. [Here](https://github.com/spxtr/test-infra/tree/jenkins/ciongke) is a direct link to the branch if GitHub or Reviewable are feeling sluggish. There are plenty of optimizations left to implement, such as deleting the old Jenkins job when the PR is updated, but I think this is a good place to make the switch.

Most of the logic happens in `cmd/hook`. One purposeful omission is the "build passed/failed" comment at the end of the GCE e2e run. Also, there's no explicit whitelist, it just checks the kubernetes org. Everyone in kubernetes-maintainers and kubernetes-collaborators is in the org. Everything else should feel pretty much the same. It doesn't quite support multiple repositories yet but it won't be hard to make it do so.

I'll add some more tests and document some things better before checking in, but it's overall ready for review. cc @krousey 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/test-infra/453)
<!-- Reviewable:end -->

rmmh commented:
Reviewed 16 of 379 files at r1, 3 of 81 files at r2.
Review status: 19 of 459 files reviewed at latest revision, 15 unresolved discussions.

---

*[ciongke/Makefile, line 25 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QHjKUly5hz0ibSkz:-KQ7QHjKUly5hz0ibSl-:brtmaae) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L25)):*
> ```Makefile
> # token of whatever account you want to comment and update statuses.
> # The Jenkins secret is the API token.
> HOOK_SECRET_FILE = $(shell echo $${HOME}/hook)
> ```

why does ${HOME}/hook not work?

---

*[ciongke/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QfFN3-JQRfL86suW:-KQ7QfFOEnH8mpAFN6fV:b-ct5cet) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L47)):*
> ```Makefile
> 	make hook-service --no-print-directory
> 	@echo -n "Waiting for loadbalancer ingress "; while [[ "$$(kubectl get svc hook -o=json | jq -r .status.loadBalancer.ingress[0].ip)" == "null" ]]; do echo -n "."; sleep 5; done; echo " Done"
> 	@echo "Webhook address: http://$$(kubectl get svc hook -o=json | jq -r .status.loadBalancer.ingress[0].ip):8888/"
> ```

This is going to be stored as a DNS entry, right?

---

*[ciongke/cmd/hook/githubagent.go, line 37 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7R6YHOZSot-qXVrtZ:-KQ7R6YHOZSot-qXVrt_:bty84f1) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L37)):*
> ```Go
> 	JenkinsJobs []JenkinsJob
> 
> 	PullRequestEvents  chan github.PullRequestEvent
> ```

Please specify the channel direction here (`<-chan` or `chan<-`), so it's clearer what direction the events are flowing.

---

*[ciongke/cmd/hook/githubagent.go, line 56 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7SpcQHuXLZoFY9Frv:-KQ7SpcRmJ6Kp7SQVJ31:bbysuxi) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L56)):*
> ```Go
> type BuildRequest struct {
> 	// If true, create a new job after deleting the old.
> 	Create bool
> ```

This is a bit confusing. Would some sort of string or enum stating what the desired action is ("run", "delete") make more sense?

---

*[ciongke/cmd/hook/githubagent.go, line 59 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7RIYlqJmHz1xhkeMI:-KQ7RIYmVM2s9ggrrmjn:b-m5xnei) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L59)):*
> ```Go
> 
> 	JobName string
> 	Context string
> ```

What is this? Most of these fields could use comments giving examples of what their values are.

---

*[ciongke/cmd/hook/githubagent.go, line 61 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7RjERo8pDFGjhT-QT:-KQ7RjERo8pDFGjhT-QU:b-rd8rq5) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L61)):*
> ```Go
> 	Context string
> 
> 	RepoOwner string
> ```

What needs these to be split up? Treating them as a single `Repo: "kubernetes/kubernetes"` seems easier.

---

*[ciongke/cmd/hook/githubagent.go, line 261 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7SgfI1bIQBvkYVYBA:-KQ7SgfI1bIQBvkYVYBB:b-nllun0) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L261)):*
> ```Go
> func (ga *GitHubAgent) deleteAll(pr github.PullRequest) {
> 	for _, job := range ga.JenkinsJobs {
> 		if !job.AlwaysRun {
> ```

Why not delete manually triggered non-default jobs on PR close?

---

*[ciongke/cmd/hook/githubagent_test.go, line 27 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7TowIcpP2wTr1tMYR:-KQ7TowIcpP2wTr1tMYS:bmu6nby) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent_test.go#L27)):*
> ```Go
> )
> 
> func TestTrusted(t *testing.T) {
> ```

What happens if an org member comments "ok to test" and Evil User pushes a new commit?

---

*[ciongke/cmd/hook/githubagent_test.go, line 67 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7U4OQJ_7EWhjz4HqD:-KQ7U4OQJ_7EWhjz4HqE:b72o0bb) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent_test.go#L67)):*
> ```Go
> 			Comments: []github.IssueComment{
> 				{
> 					Body: "ok to test",
> ```

test that "not ok to test" doesn't match

---

*[ciongke/cmd/hook/githubagent_test.go, line 253 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7UY_zkE1IMyv2oLJo:-KQ7UY_zkE1IMyv2oLJp:b-k4oftd) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent_test.go#L253)):*
> ```Go
> 		}
> 	}
> }
> ```

This needs a couple more tests:

- triggering specific builds
- closing a PR canceling builds in progress

---

*[ciongke/cmd/hook/kubeagent.go, line 35 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7VdKqScIpM58MlCUo:-KQ7VdKrr6OJuE88IEU3:b-arkjhv) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L35)):*
> ```Go
> 	Namespace   string
> 
> 	BuildRequests chan BuildRequest
> ```

Put channel direction in type.

---

*[ciongke/cmd/hook/kubeagent.go, line 66 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7YMqD8LYh6P8G2-pO:-KQ7YMqD8LYh6P8G2-pP:b1ylh4x) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L66)):*
> ```Go
> 
> func (ka *KubeAgent) createJob(br BuildRequest) error {
> 	name := fmt.Sprintf("%s-pr-%d-%s", br.RepoName, br.PR, br.JobName)
> ```

Include RepoOwner in job name.

---

*[ciongke/cmd/hook/kubeagent.go, line 72 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7YTi9qqYN1AIKP7Tk:-KQ7YTi9qqYN1AIKP7Tl:b-9qz1i3) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L72)):*
> ```Go
> 			Namespace: ka.Namespace,
> 			Labels: map[string]string{
> 				"repo":             br.RepoName,
> ```

put RepoOwner here (or with a different label)

---

*[ciongke/cmd/hook/server_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7_KdGjlDytWyhlrIt:-KQ7_KdGjlDytWyhlrIu:b-sxj8f5) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/server_test.go#L42)):*
> ```Go
> 		{
> 			// GET
> 			Method: http.MethodGet,
> ```

I'm really ok with 'GET' and 'POST' and 403 instead of these constants, especially since that's what you use in the other file.

---

*[ciongke/cmd/test-pr/main.go, line 41 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7aA5N_tsuc7NtPkVz:-KQ7aA5N_tsuc7NtPkW-:b-7eal2p) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/test-pr/main.go#L41)):*
> ```Go
> 	dryRun    = flag.Bool("dry-run", true, "Whether or not to make mutating GitHub/Jenkins calls.")
> 
> 	githubTokenFile  = flag.String("github-token-file", "/etc/oauth/oauth", "Path to the file containing the GitHub OAuth secret.")
> ```

`oauth/oauth` isn't very descriptive. How about `/etc/pr/github-token` and `/etc/pr/jenkins-token`, etc?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

spxtr commented:
Review status: 19 of 459 files reviewed at latest revision, 16 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7uHhF8P7deQmumwVF:-KQ7uHhF8P7deQmumwVG:b-jq092x) (no related file):*
Thanks. I need to clarify the logic in a couple places, I'll ping you when it's ready for re-review.

---

*[ciongke/Makefile, line 25 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QHjKUly5hz0ibSkz:-KQ7eSJ9L4qXW27uuCG1:bf1g076) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L25)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> why does ${HOME}/hook not work?
</details>
It does, and I don't know why I did this. Fixed.

---

*[ciongke/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QfFN3-JQRfL86suW:-KQ7ee7qyWM7LvDD5fzC:b-92jrd6) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L47)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> This is going to be stored as a DNS entry, right?
</details>
No, it's the raw IP address that we point github at.

---

*[ciongke/cmd/hook/githubagent.go, line 37 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7R6YHOZSot-qXVrtZ:-KQ7i5AW49foNV_b2TUv:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L37)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> Please specify the channel direction here (`<-chan` or `chan<-`), so it's clearer what direction the events are flowing.
</details>
Done.

---

*[ciongke/cmd/hook/githubagent.go, line 56 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7SpcQHuXLZoFY9Frv:-KQ7itw2bQepgPpBlEU2:bplylbu) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L56)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> This is a bit confusing. Would some sort of string or enum stating what the desired action is ("run", "delete") make more sense?
</details>
That might be a good idea. It's also poorly named, the default value for "BuildRequest" actually only deletes! Maybe renaming this to "KubeRequest" would be good?

---

*[ciongke/cmd/hook/githubagent.go, line 59 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7RIYlqJmHz1xhkeMI:-KQ7iipH_mOjeezIsuA1:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L59)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> What is this? Most of these fields could use comments giving examples of what their values are.
</details>
Done.

---

*[ciongke/cmd/hook/githubagent.go, line 61 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7RjERo8pDFGjhT-QT:-KQ7iP8wBMifvjv64mOH:b-umo7xe) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L61)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> What needs these to be split up? Treating them as a single `Repo: "kubernetes/kubernetes"` seems easier.
</details>
The github client needs them to be separate.

---

*[ciongke/cmd/hook/githubagent.go, line 261 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7SgfI1bIQBvkYVYBA:-KQ7jT2t6KnSUy1mc0Jz:bt5sz33) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L261)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> Why not delete manually triggered non-default jobs on PR close?
</details>
Good catch. I need to clarify the logic around this and add a test. Will do.

---

*[ciongke/cmd/hook/githubagent_test.go, line 27 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7TowIcpP2wTr1tMYR:-KQ7jjlyMeilEBXii1zX:byvi7lw) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent_test.go#L27)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> What happens if an org member comments "ok to test" and Evil User pushes a new commit?
</details>
Then we build it. That's the way it is now. We can consider dropping this functionality and only allowing "@k8s-bot test this".

---

*[ciongke/cmd/hook/githubagent_test.go, line 67 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7U4OQJ_7EWhjz4HqD:-KQ7ln3uYDs5Q1jHBb5a:bwoorwl) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent_test.go#L67)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> test that "not ok to test" doesn't match
</details>
Done. Not sure how it currently works with the PR plugin, but I made it match `(?m)^ok to test$`.

---

*[ciongke/cmd/hook/githubagent_test.go, line 253 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7UY_zkE1IMyv2oLJo:-KQ7qQO1I9FUmTUZ2ZHF:b-sglbwe) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent_test.go#L253)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> This needs a couple more tests:
> 
> - triggering specific builds
> - closing a PR canceling builds in progress
</details>
1. Done.
2. I need to go over that logic. Will fix.

---

*[ciongke/cmd/hook/kubeagent.go, line 35 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7VdKqScIpM58MlCUo:-KQ7qZyPfhw-hJUaiZho:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L35)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> Put channel direction in type.
</details>
Done.

---

*[ciongke/cmd/hook/kubeagent.go, line 66 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7YMqD8LYh6P8G2-pO:-KQ7rKcwWTG6L5JcZ6A1:bktp9ct) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L66)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> Include RepoOwner in job name.
</details>
Done, although I'll have to check if there's an issue with long object names. Might have to name it with a uuid or something.

---

*[ciongke/cmd/hook/kubeagent.go, line 72 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7YTi9qqYN1AIKP7Tk:-KQ7rV_arlJyPLg-Nv9n:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L72)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> put RepoOwner here (or with a different label)
</details>
Done.

---

*[ciongke/cmd/test-pr/main.go, line 41 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7aA5N_tsuc7NtPkVz:-KQ7tijsVoskyj9YFCwS:b17717a) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/test-pr/main.go#L41)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> `oauth/oauth` isn't very descriptive. How about `/etc/pr/github-token` and `/etc/pr/jenkins-token`, etc?
</details>
Changed it to `/etc/webhook/hmac` and `/etc/github/oauth`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

rmmh commented:
Reviewed 4 of 379 files at r1, 9 of 9 files at r3.
Review status: 24 of 459 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[ciongke/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QfFN3-JQRfL86suW:-KQ8B4Su1vMjsORylkZ9:b-qft5vy) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L47)):*
<details><summary><i>Previously, spxtr (Joe Finney) wrote…</i></summary>
> No, it's the raw IP address that we point github at.
</details>
If this isn't using DNS, make sure it's using a static IP for the ingress IP. Losing webhooks because the IP moved would be unfortunate.

---

*[ciongke/cmd/test-pr/main.go, line 141 \[r3\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ8Ij1dZqNObxHlZrCP:-KQ8Ij1edkIaE8u4okYI:b-ei65rb) ([raw file](https://github.com/kubernetes/test-infra/blob/db46ab60cff47d8cdb7bc9ec20e3eef6b8692a05/ciongke/cmd/test-pr/main.go#L141)):*
> ```Go
> 		result, err := c.JenkinsClient.Status(b)
> 		if err != nil {
> 			c.tryCreateStatus(github.Error, "Error waiting for build.", "")
> ```

We should revisit this in the future-- occasional network hiccups shouldn't affect this.

---

*[ciongke/cmd/test-pr/main.go, line 151 \[r3\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ8Hh-MA_YW8LNaW6S8:-KQ8Hh-MA_YW8LNaW6S9:b-50j93s) ([raw file](https://github.com/kubernetes/test-infra/blob/db46ab60cff47d8cdb7bc9ec20e3eef6b8692a05/ciongke/cmd/test-pr/main.go#L151)):*
> ```Go
> 				break
> 			} else {
> 				c.tryCreateStatus(github.Failure, "Build failed.", result.URL)
> ```

Can you give more details in the failure message -- like distinguishing between timeouts and failures?

---

*[ciongke/jenkins/jenkins.go, line 164 \[r3\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ8KjVeqoBpC-X9h14X:-KQ8KjVeqoBpC-X9h14Y:b-z9lbud) ([raw file](https://github.com/kubernetes/test-infra/blob/db46ab60cff47d8cdb7bc9ec20e3eef6b8692a05/ciongke/jenkins/jenkins.go#L164)):*
> ```Go
> 		}, nil
> 	}
> 	u := fmt.Sprintf("%s/job/%s/api/json?depth=1", c.baseURL, b.jobName)
> ```

You can use "tree" to limit the results to just what you want, instead of depth=1:

`/json?tree=builds[number,result,actions[parameters[name,value]]]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

rmmh commented:
Reviewed 1 of 379 files at r1.
Review status: 25 of 459 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

rmmh commented:
Reviewed 13 of 18 files at r4.
Review status: 24 of 463 files reviewed at latest revision, 20 unresolved discussions.

---

*[ciongke/cmd/hook/kubeagent.go, line 86 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQN5sy76hRT62sPTSEW:-KQN5sy76hRT62sPTSEX:b-ne6wje) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L86)):*
> ```Go
> 
> func (ka *KubeAgent) createJob(kr KubeRequest) error {
> 	name := uuid.NewV1().String()
> ```

Don't many dashboards show just the job name? Having a UUID is a bit harder to debug. Kubernetes seems to do base + 5-char random string for pod names. If there's a specific length limit, it would be easy to truncate when necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

spxtr commented:
Review status: 24 of 463 files reviewed at latest revision, 20 unresolved discussions.

---

*[ciongke/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QfFN3-JQRfL86suW:-KQHupgFQzp9RiE2mckz:bm5j5jc) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L47)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> If this isn't using DNS, make sure it's using a static IP for the ingress IP. Losing webhooks because the IP moved would be unfortunate.
</details>
Oh good catch, I need to bump the ephemeral IP to a static IP. Done.

---

*[ciongke/cmd/hook/githubagent.go, line 56 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7SpcQHuXLZoFY9Frv:-KQMWy2Astpo-4cMicea:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/githubagent.go#L56)):*
<details><summary><i>Previously, spxtr (Joe Finney) wrote…</i></summary>
> That might be a good idea. It's also poorly named, the default value for "BuildRequest" actually only deletes! Maybe renaming this to "KubeRequest" would be good?
</details>
Done.

---

*[ciongke/cmd/hook/kubeagent.go, line 66 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7YMqD8LYh6P8G2-pO:-KQ8KWHheJPmjbQqF6F0:b-iwrp8a) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/kubeagent.go#L66)):*
<details><summary><i>Previously, spxtr (Joe Finney) wrote…</i></summary>
> Done, although I'll have to check if there's an issue with long object names. Might have to name it with a uuid or something.
</details>
Yep, limit of 63 characters in the job name. 

---

*[ciongke/cmd/hook/kubeagent.go, line 86 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQN5sy76hRT62sPTSEW:-KQN8Fm-YAV8DmPPy1gc:b-hwmnv7) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L86)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> Don't many dashboards show just the job name? Having a UUID is a bit harder to debug. Kubernetes seems to do base + 5-char random string for pod names. If there's a specific length limit, it would be easy to truncate when necessary.
</details>
The standard kubernetes dashboard shows the labels as well. The length limit is 63 characters, which is exactly the length of "kubernetes-kubernetes-pr-12345-kubernetes-pull-build-test-e2e-g", so naive truncating won't work.

---

*[ciongke/cmd/hook/server_test.go, line 42 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7_KdGjlDytWyhlrIt:-KQHvBp6uGZmyWAwH46U:bb6sh2i) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/cmd/hook/server_test.go#L42)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> I'm really ok with 'GET' and 'POST' and 403 instead of these constants, especially since that's what you use in the other file.
</details>
Ack.

---

*[ciongke/cmd/test-pr/main.go, line 141 \[r3\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ8Ij1dZqNObxHlZrCP:-KQHvE0jqWlqUwHMG1HZ:ba983fn) ([raw file](https://github.com/kubernetes/test-infra/blob/db46ab60cff47d8cdb7bc9ec20e3eef6b8692a05/ciongke/cmd/test-pr/main.go#L141)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> We should revisit this in the future-- occasional network hiccups shouldn't affect this.
</details>
Agreed, and we should probably have retries built into the kube/jenkins/github clients. I think that can happen in another PR.

---

*[ciongke/cmd/test-pr/main.go, line 151 \[r3\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ8Hh-MA_YW8LNaW6S8:-KQMTjF3T2kjeaCRxg0t:b-2iv6yb) ([raw file](https://github.com/kubernetes/test-infra/blob/db46ab60cff47d8cdb7bc9ec20e3eef6b8692a05/ciongke/cmd/test-pr/main.go#L151)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> Can you give more details in the failure message -- like distinguishing between timeouts and failures?
</details>
Yes, I'd love to do that in another PR. There's the link to gubernator for now.

---

*[ciongke/jenkins/jenkins.go, line 164 \[r3\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ8KjVeqoBpC-X9h14X:-KQN88caKdtCTBWvp4aU:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/db46ab60cff47d8cdb7bc9ec20e3eef6b8692a05/ciongke/jenkins/jenkins.go#L164)):*
<details><summary><i>Previously, rmmh (Ryan Hitchman) wrote…</i></summary>
> You can use "tree" to limit the results to just what you want, instead of depth=1:
> 
> `/json?tree=builds[number,result,actions[parameters[name,value]]]`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

krousey commented:
Reviewed 357 of 379 files at r1, 78 of 81 files at r2, 5 of 18 files at r4.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[ciongke/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QfFN3-JQRfL86suW:-KQRPJdmPL0z6ro1niJV:b-ktsx8v) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L47)):*
<details><summary><i>Previously, spxtr (Joe Finney) wrote…</i></summary>
> Oh good catch, I need to bump the ephemeral IP to a static IP. Done.
</details>
Is there a reason to not use DNS?

---

*[ciongke/cmd/hook/kubeagent.go, line 63 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRIW7yjuQCEwZkJ31h:-KQRIW7z1HGutyPy1Q7B:bakw4vg) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L63)):*
> ```Go
> 
> // Cut off test-pr jobs after 10 hours.
> const jobDeadlineSeconds = 60 * 60 * 10
> ```

`const jobDeadline = 10 * time.Hour`

---

*[ciongke/cmd/hook/kubeagent.go, line 65 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRL4JfNbX-r6sv5pWk:-KQRL4JgtVHL3NDq4ao_:b50pq0m) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L65)):*
> ```Go
> const jobDeadlineSeconds = 60 * 60 * 10
> 
> func (ka *KubeAgent) Start() {
> ```

How do you stop this thing?

---

*[ciongke/cmd/hook/kubeagent.go, line 99 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRKDQ8VL9P6I_UKcao:-KQRKDQ9D5bbhIsB0tpT:b-7l9xty) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L99)):*
> ```Go
> 		},
> 		Spec: kube.JobSpec{
> 			ActiveDeadlineSeconds: jobDeadlineSeconds,
> ```

`int64(jobDeadline / time.Second)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

spxtr commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[ciongke/Makefile, line 47 \[r1\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQ7QfFN3-JQRfL86suW:-KQRZOJLRgCd0yI9YQLf:b-uusmod) ([raw file](https://github.com/kubernetes/test-infra/blob/ab73135de731290db0c3d65da1f15f66033a1d12/ciongke/Makefile#L47)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> Is there a reason to not use DNS?
</details>
I haven't used Cloud DNS before, but other than that there's no reason.

---

*[ciongke/cmd/hook/kubeagent.go, line 65 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRL4JfNbX-r6sv5pWk:-KQRZs87JImcwk011JWg:b-e0qimb) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L65)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> How do you stop this thing?
</details>
In main.go I ignore term, which means it waits 30 seconds and then gets killed. I'm assuming all requests take <30s to finish.

I could add a stop channel and have main.go wait some amount of time after getting term before stopping them but I'm not sure it's worth it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

spxtr commented:
Review status: 461 of 463 files reviewed at latest revision, 23 unresolved discussions.

---

*[ciongke/cmd/hook/kubeagent.go, line 63 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRIW7yjuQCEwZkJ31h:-KQRg9JUj27oHPlh6YJm:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L63)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> `const jobDeadline = 10 * time.Hour`
</details>
Done.

---

*[ciongke/cmd/hook/kubeagent.go, line 99 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRKDQ8VL9P6I_UKcao:-KQRgPd90rKXrpk2NHT1:b-896fix) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L99)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> `int64(jobDeadline / time.Second)`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

spxtr commented:
Review status: 461 of 463 files reviewed at latest revision, 23 unresolved discussions.

---

*[ciongke/cmd/hook/kubeagent.go, line 65 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRL4JfNbX-r6sv5pWk:-KQS3SvXbC0pORScvgUI:b-uv3ube) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L65)):*
<details><summary><i>Previously, spxtr (Joe Finney) wrote…</i></summary>
> In main.go I ignore term, which means it waits 30 seconds and then gets killed. I'm assuming all requests take <30s to finish.
> 
> I could add a stop channel and have main.go wait some amount of time after getting term before stopping them but I'm not sure it's worth it.
</details>
I made it use a range rather than select so if they get closed for some reason it won't spin.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

krousey commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[ciongke/cmd/hook/kubeagent.go, line 65 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRL4JfNbX-r6sv5pWk:-KQS7mg-pesg1pt-dKHE:bagqzf2) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L65)):*
<details><summary><i>Previously, spxtr (Joe Finney) wrote…</i></summary>
> I made it use a range rather than select so if they get closed for some reason it won't spin.
</details>
Are you concerned at all about a race between a build request and a delete?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

spxtr commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[ciongke/cmd/hook/kubeagent.go, line 65 \[r4\]](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-KQRL4JfNbX-r6sv5pWk:-KQSAkIIzrer4OzXZxtS:bjh46hh) ([raw file](https://github.com/kubernetes/test-infra/blob/7f03615da8c7b999a0b682e5e82fe08b02b5cf76/ciongke/cmd/hook/kubeagent.go#L65)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> Are you concerned at all about a race between a build request and a delete?
</details>
No, I never put something on the delete channel and then the build channel. Only one or the other.

It's possible that it races with one of the other hook pods if, for instance, you open a PR and comment "test this" immediately. This might result in a build request on each hook pod, and because kubeagent's delete then build isn't atomic we might end up with two Jenkins jobs. I don't think that's the end of the world, since it does that now anyway :P

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453)*
<!-- Sent from Reviewable.io -->

rmmh commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 356 of 379 files at r1, 78 of 81 files at r2, 4 of 18 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-:-KQT2z7bcCgBEiAHLS4w:bnfp4nl)*
<!-- Sent from Reviewable.io -->

krousey commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/test-infra/453#-:-KQXQ2O4uY_RPgzOAnFV:bnfp4nl)*
<!-- Sent from Reviewable.io -->
